dpnp/dpnp_iface_linearalgebra.py

```diff
@@ -269,63 +330,325 @@ def matmul(x1, x2, out=None, **kwargs):
+
+        if list_of_events:
+            for ev in list_of_events:
+                ev.wait()
+
+        # calculate results
+        result = dpnp.empty(
+            res_shape,
+            dtype=gemm_dtype,
+            usm_type=res_usm_type,
+            sycl_queue=exec_q,
+        )
+        if result.size == 0:
             pass
         else:
-            if 0:
-                """
-                Cost model checks
-                """
-
-                array1_size = x1_desc.size
-                array2_size = x2_desc.size
-                cost_size = 4096  # 2D array shape(64, 64)
-
-                if (x1_desc.dtype == dpnp.float64) or (
-                    x1_desc.dtype == dpnp.float32
-                ):
-                    """
-                    Floating point types are handled via original math library better than SYCL math library
-                    """
-                    cost_size = 262144  # 2D array shape(512, 512)
-
-                if (array1_size > cost_size) and (array2_size > cost_size):
-                    return dpnp_matmul(x1_desc, x2_desc, out)
+            if x1.size == 0 or x2.size == 0:
+                result = dpnp.zeros(
+                    res_shape,
+                    dtype=gemm_dtype,
+                    usm_type=res_usm_type,
+                    sycl_queue=exec_q,
+                )
             else:
-                out_desc = (
-                    dpnp.get_dpnp_descriptor(
-                        out, copy_when_nondefault_queue=False
+                ht_copy_ev_x1 = dpctl.SyclEvent()
+                ht_copy_ev_x2 = dpctl.SyclEvent()
+                if x1_ndim == 2 and x2_ndim == 2:
+                    ht_blas_ev, _ = bi._gemm(
+                        exec_q,
+                        dpnp.get_usm_ndarray(x1),
+                        dpnp.get_usm_ndarray(x2),
+                        dpnp.get_usm_ndarray(result),
+                        [],
                     )
-                    if out is not None
-                    else None
-                )
-                return dpnp_matmul(x1_desc, x2_desc, out_desc).get_pyobj()
+                else:
+                    (
+                        ht_blas_ev,
+                        ht_copy_ev_x1,
+                        ht_copy_ev_x2,
+                        result,
+                    ) = dpnp_matmul_batch(
+                        exec_q,
+                        x1,
+                        x2,
+                        result,
+                        x1_is_2D,
+                        x2_is_2D,
+                        ht_copy_ev_x1,
+                        ht_copy_ev_x2,
+                    )
+
+                ht_blas_ev.wait()
+                ht_copy_ev_x1.wait()
+                ht_copy_ev_x2.wait()
+
+        if squeeze_flag:
+            result = dpnp.squeeze(result)
+
+        if gemm_dtype != res_dtype:
+            result = dpnp.astype(result, res_dtype)
+
+        if out is not None:
```

Comment:
Since we only perform the calculations for `matmul` in floating point data type using OneMKL. If the input is integer type, the intermediate result would be in floating point and if the output from out keyword is integer we will get an error unless we have change the data type of intermediate result to integer.

Reply:
but then we should do that only with `copy=False` and for non floating types - to avoid extra copying

Reply:
Yes, I have added `copy=False` in recent updates and the `if gemm_dtype != res_dtype` will only be true for for non-floating types.